com.unity.shadergraph/Editor/Generation/IHasMetaData.cs

```diff
@@ -6,6 +6,6 @@ namespace UnityEditor.ShaderGraph
     interface IHasMetadata
     {
         string identifier { get; }
-        ScriptableObject GetMetadataObject();
+        ScriptableObject GetMetadataObject(GraphData graph);
```

Comment:
This interface seems really broad reaching. Modifiable access to the entire graph object is a drastic change in the accessing pattern here.

Reply:
It's true, even if IHasMetadata isn't a public interface, it can be really unexpected to modify the GraphData while calling GetMetadataObject, especially since it's done in importer pass.

Some context, I'm using it to store an information about the shaderGraph behavior in HDRP https://github.com/Unity-Technologies/Graphics/blob/a56b36b34da286367fb192088f903b8523a29811/com.unity.render-pipelines.high-definition/Editor/Material/ShaderGraph/HDSubTarget.cs#L83
In brief, if there is a vertex modification, we can't use the optimized path from VFX because we aren't applying code from SG in compute shader.

At some point, we would like to remove the ShaderGraphVfxAsset and only relies on MetaData, so, we would expect to store some attributes only readable in GraphData.

I can potentially replace it with a `struct ReadOnlyGraphData` which will be limited to our today's need, here a draft :
```
//GraphDataReadOnly main contract is "doesn't modify the encapsulated GraphData"
struct GraphDataReadOnly
{
   private GraphData m_GraphData
   GraphDataReadOnly(GraphData graphData)
   {
      m_GraphData = graphData;
   }

   public bool AnyConnectedControl<T>() where T : IControl
   {
       var matchingNodes = graphData.GetNodes<BlockNode>().Where(o => o.descriptor.control is T);
       return matchingNodes.SelectMany(o => o.GetInputSlots<MaterialSlot>()).Any(o => o.isConnected));
   }
}
```
This way, we have a read graph data helper we can use in GetMetadata, we could extend this structure with other read helper (insuring we are always returning immutable data).

Thus, I'm replacing 
```
ScriptableObject GetMetadataObject(GraphData graph);
``` 
by 
```
ScriptableObject GetMetadataObject(GraphDataReadOnly graph);
```

What do you think about this approach ? I am open to any alternative. 

I also thought about some option I elimated :
- An interface directly on GraphData but it's too easy to cast from an interface to the base type in C#
- Serialize/Deserialize the GraphData between each GetMetaData call, it would be hacky and really expensive.

Reply:
This seems like a decent approach...  the class basically represents the externally queryable data about the graph.  You could make it even more restrictive for now, and have the public function be:

public bool AnyVertexAnimationActive() { ... }

Reply:
which will at least tide us over til we figure out how to handle these types of queries more generically in the future...

Reply:
I've also seen it done so that nodes in the graph can produce queryable data... though in this case you're specifically looking for whether the vertex position input is defaulted or not, right?

Reply:
Fixed at 969b89ed88dbfef49fb9d2d19b8cfbae336af415